pkg/placement/hashing/consistent_hash.go

```diff
@@ -90,7 +82,104 @@ func NewConsistentHash() *Consistent {
 }
 
 // NewFromExisting creates a new consistent hash from existing values.
-func NewFromExisting(hosts map[uint64]string, sortedSet []uint64, loadMap map[string]*Host) *Consistent {
+func NewFromExisting(loadMap map[string]*Host, replicationFactor int, virtualNodesCache *VirtualNodesCache) *Consistent {
+	newHash := &Consistent{
+		hosts:     map[uint64]string{},
+		sortedSet: []uint64{},
+		loadMap:   loadMap,
+	}
+
+	for hostName := range loadMap {
+		hashes := virtualNodesCache.GetHashes(int64(replicationFactor), hostName)
+		for _, h := range hashes {
+			newHash.hosts[h] = hostName
+		}
+		newHash.sortedSet = append(newHash.sortedSet, hashes...)
+	}
+
+	// sort hashes in ascending order
+	sort.Slice(newHash.sortedSet, func(i int, j int) bool {
+		return newHash.sortedSet[i] < newHash.sortedSet[j]
+	})
+
+	return newHash
+}
+
+// VirtualNodesCache data example:
+//
+//	100 -> (the replication factor)
+//		"192.168.1.89:62362" -> ["10056481384176189962", "10100244799470048543"...] (100 elements)
+//		"192.168.1.89:62362" -> ["10056481384176189962", "10100244799470048543"...]
+//	200 ->
+//		"192.168.1.89:62362" -> ["10056481384176189962", "10100244799470048543"...] (200 elements)
+//		"192.168.1.89:62362" -> ["10056481384176189962", "10100244799470048543"...]
+type VirtualNodesCache struct {
+	sync.RWMutex
+	data map[int64]*hashMap
+}
+
+// hashMap represents a mapping of IP addresses to their hashes.
+type hashMap struct {
+	hashes map[string][]uint64
+}
+
+func newHashMap() *hashMap {
+	return &hashMap{
+		hashes: make(map[string][]uint64),
+	}
+}
+
+func NewVirtualNodesCache() *VirtualNodesCache {
+	return &VirtualNodesCache{
+		data: make(map[int64]*hashMap),
+	}
+}
+
+// GetHashes retrieves the hashes for the given replication factor and IP address.
+func (hc *VirtualNodesCache) GetHashes(replicationFactor int64, host string) []uint64 {
+	hc.RLock()
+	if hashMap, exists := hc.data[replicationFactor]; exists {
+		if hashes, found := hashMap.hashes[host]; found {
+			return hashes
```

Comment:
Don't we need `hc.RUnlock()` before this?

Reply:
What a save! Thank you.